server/server/src/billing/billing_service.rs

```diff
@@ -379,113 +514,300 @@ pub async fn stripe_webhooks(
-                        "Cannot retrieve the customer_id.".to_string(),
-                    ))
-                })? {
-                    stripe::Expandable::Id(id) => id.to_string(),
-                    stripe::Expandable::Object(customer) => customer.id.to_string(),
-                };
-
-                let (public_key, mut user_info) =
-                    get_public_key_and_stripe_user_info(&event, &mut con, &customer_id).await?;
-                let pos = get_active_subscription_index(&user_info.subscriptions)?;
+                let tx_result = tx!(&mut con, pipe, &[keys::subscription_profile(&public_key)], {
+                    match lock_billing_workflow(
+                        &public_key,
+                        &mut con,
+                        server_state.config.stripe.millis_between_user_payment_flows,
+                    )
+                    .await
+                    {
+                        Ok(ref mut sub_profile) => {
+                            if let Some(BillingPlatform::Stripe(ref mut info)) =
+                                sub_profile.billing_platform
+                            {
+                                info.expiration_time = subscription_period_end as u64;
+                            }
+
+                            sub_profile.last_in_payment_flow = 0;
+                            pipe.json_set(keys::subscription_profile(&public_key), &sub_profile)
+                        }
+                        Err(ClientError(LockBillingWorkflowError::ExistingRequestPending)) => {
+                            Ok(&mut pipe)
+                        }
+                        Err(err) => Err(Abort(internal!(
+                            "Cannot get billing lock in stripe webhooks: {:#?}",
+                            err
+                        ))),
+                    }
+                });
+                return_if_error!(tx_result);
 
                 info!(
-                    "User's subscription period_end is being changed after successful renewal. public_key: {}",
+                    "User's subscription period_end is changed after successful renewal. public_key: {}",
                     keys::stringify_public_key(&public_key)
                 );
-
-                user_info.subscriptions[pos].period_end = subscription_period_end as u64;
-
-                con.json_set(stripe_user_info(&public_key), &user_info)
-                    .await?;
             }
         }
         (_, _) => {
-            return Err(internal!("Unexpected and unhandled stripe event: {:?}", event.event_type))
+            return Err(internal!("Unexpected and unhandled stripe event: {:?}", event.event_type));
         }
     }
 
     Ok(())
 }
 
-async fn get_public_key_and_stripe_user_info(
-    event: &stripe::WebhookEvent, con: &mut Connection, customer_id: &str,
-) -> Result<(PublicKey, StripeUserInfo), ServerError<StripeWebhookError>> {
+async fn get_public_key(
```

Comment:
this fn's name was better before lol. I think you may want to have higher level stripe stuff in it's own file, and higher level play stuff in it's own file. Then billing_service would just be an implementation of the document I was referring to before and you could very clearly trace what is allowed and what is not allowed.